besskge/scoring.py

```diff
@@ -602,6 +638,254 @@ def score_tails(
+            [embedding_size, embedding_size],
+        )
+        # self.relation_embedding[..., :embedding_size] head box centers
+        # self.relation_embedding[..., embedding_size:2*embedding_size] tail box centers
+        # self.relation_embedding[..., 2*embedding_size:3*embedding_size] head box widths
+        # self.relation_embedding[..., 3*embedding_size:4*embedding_size] tail box widths
+        # self.relation_embedding[..., -2] head box size
+        # self.relation_embedding[..., -1] tail box size
+        self.relation_embedding = initialize_relation_embedding(
+            n_relation_type,
+            4 * [relation_initializer[0]] + 2 * [relation_initializer[1]]
+            if isinstance(relation_initializer, list)
+            else relation_initializer,
+            [embedding_size, embedding_size, embedding_size, embedding_size, 1, 1],
+        )
+        assert (
+            2 * self.entity_embedding.shape[-1]
+            == self.relation_embedding.shape[-1] - 2
+            == 4 * embedding_size
+        ), (
+            "BoxE requires `2*embedding_size` embedding parameters for each entity"
+            " and `4*embedding_size + 2` embedding parameters for each relation"
+        )
+        self.embedding_size = embedding_size
+
+    def boxe_score(
+        self,
+        center_dist_ht: torch.Tensor,
+        width_ht: torch.Tensor,
+        box_size: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        BoxE score with relation broadcasting, for optimized batched negative scoring.
+
+        :param center_dist_ht: shape: (batch_size, 2, emb_size)
+            or (batch_size, n_negative, 2, emb_size)
+            Distances of bumped h/t entity embeddings from center of h/t relation boxes
+            (heads: `center_dist_ht[...,0,:]`, tails: `center_dist_ht[...,1,:]`).
+        :param width_ht: shape: (batch_size, 2, emb_size) or (batch_size, 1, 2, emb_size)
+            Widths of h/t relation boxes, before normalization
+            (heads: `width_ht[...,0,:]`, tails: `width_ht[...,1,:]`).
+        :param box_size: shape: (batch_size, 2) or (batch_size, 1, 2)
+            Parameter controlling the size of the (normalized)
+            h/t relation boxes
+            (heads: `box_size[...,0]`, tails: `box_size[...,1]`).
+
+        :return: shape: shape: (batch_size,) or (batch_size, n_negative)
+            Negative sum of head and tail scores, as defined in :cite:p:`BoxE`.
+        """
+        width_ht = torch.abs(width_ht)
+        # Normalize width_ht by geometric mean
+        width_ht = width_ht / torch.clamp(
+            torch.exp(
+                torch.mean(
+                    torch.log(torch.clamp(width_ht, min=self.soft)),
+                    dim=-1,
+                    keepdim=True,
+                )
+            ),
+            min=self.soft,
+        )
+        # Rescale by box_size
+        width_ht = width_ht * (
+            torch.tensor(1.0, dtype=torch.float32, device=width_ht.device)
```

Comment:
is it necessary to explicitly create these tensors (since it's just single floats)? Won't `torch.nn.functional.elu(box_size.unsqueeze(-1) + 1` perform identically?

Reply:
Ah, of course that makes sense - I'm not aware of a better way, I just ignore those warnings... Thanks 👍